include/oneapi/dpl/pstl/algorithm_impl.h

```diff
@@ -4289,6 +4289,88 @@ __pattern_shift_right(_Tag __tag, _ExecutionPolicy&& __exec, _BidirectionalItera
     return __res.base();
 }
 
+template <class _ForwardIterator, class _IdxHashFunc, class _RandomAccessIterator, class _IsVector>
+void
+__brick_histogram(_ForwardIterator __first, _ForwardIterator __last, _IdxHashFunc __func,
+                  _RandomAccessIterator __histogram_first, _IsVector) noexcept
+{
+    for (; __first != __last; ++__first)
+    {
+        std::int32_t __bin = __func.get_bin(*__first);
+        if (__bin >= 0)
```

Comment:
1) Can `get_bin` return negative value?
2) If yes, what is correct behavior  for `__brick_histogram`?

Reply:
Yes, `-1` is returned when the input element does not fall within any histogram bin.  The correct behavior is to do nothing and skip this input element.
[Specification](https://github.com/uxlfoundation/oneAPI-spec/blob/main/source/elements/oneDPL/source/parallel_api/algorithms.rst)
"Input values that do not map to a defined bin are skipped silently."

I recently looked into expanding the bin helper interface to include a separate function to check bounds, and another to get the bin which assumes it is in bounds.  I thought this might provide benefit by reducing the number of branches by 1, but I saw no performance benefit from this change for CPU or GPU.  It is still something we could pursue in the future.